discovery/gossiper.go

```diff
@@ -2345,70 +2313,234 @@ func (d *AuthenticatedGossiper) fetchNodeAnn(
 	return node.NodeAnnouncement(true)
 }
 
-// sendAnnSigReliably will try to send the provided local AnnounceSignatures
-// to the remote peer, waiting for it to come online if necessary. This
-// method returns after adding the message to persistent storage, such
-// that the caller knows that the message will be delivered at one point.
-func (d *AuthenticatedGossiper) sendAnnSigReliably(
-	msg *lnwire.AnnounceSignatures, remotePeer *btcec.PublicKey) error {
-
-	// We first add this message to the database, such that in case
-	// we do not succeed in sending it to the peer, we'll fetch it
-	// from the DB next time we start, and retry.
-	var remotePubKey [33]byte
-	copy(remotePubKey[:], remotePeer.SerializeCompressed())
-	if err := d.cfg.MessageStore.AddMessage(msg, remotePubKey); err != nil {
+// sendMsgReliably attempts to send a message reliably to a peer. This is done
+// by first writing the message to disk, and then launching a background
+// goroutine that will resend any pending non-stale messages to the peer upon
+// subsequent reconnections.
+func (d *AuthenticatedGossiper) sendMsgReliably(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// We'll add the message to the store to ensure we can resend it upon
+	// restarts if the peer has yet to come back online.
+	if err := d.cfg.MessageStore.AddMessage(msg, peerPubKey); err != nil {
 		return err
 	}
 
-	// We have succeeded adding the message to the database. We now launch
-	// a goroutine that will keep on trying sending the message to the
-	// remote peer until it succeeds, or the gossiper shuts down. In case
-	// of success, the message will be removed from the database.
-	d.wg.Add(1)
-	go func() {
-		defer d.wg.Done()
-		for {
-			log.Debugf("Sending AnnounceSignatures for channel "+
-				"%v to remote peer %x", msg.ChannelID,
-				remotePeer.SerializeCompressed())
-			err := d.cfg.SendToPeer(remotePeer, msg)
-			if err == nil {
-				// Sending succeeded, we can
-				// continue the flow.
-				break
+	return d.sendMsg(msg, peerPubKey)
+}
+
+// sendMsg attempts to send a message to a peer.
+func (d *AuthenticatedGossiper) sendMsg(msg lnwire.Message,
+	compressedPeerPubKey [33]byte) error {
+
+	// If there doesn't exist a connection handler for this peer, we'll
+	// launch one now. This signals that we are not aware of the peer's
+	// connection status, so we'll request to be notified upon their
+	// connection and send the message then.
+	if _, ok := d.peerConnectionHandlers[compressedPeerPubKey]; !ok {
+		peerPubKey, err := btcec.ParsePubKey(
+			compressedPeerPubKey[:], btcec.S256(),
+		)
+		if err != nil {
+			return err
+		}
+
+		peerChan := make(chan lnpeer.Peer, 1)
+		d.peerConnectionHandlers[compressedPeerPubKey] = peerChan
```

Comment:
Since we can process messages in parallel, we need a mutex here.

Reply:
Fixed - now uses the `cachedPeersMtx`.